ansible/roles/azure/tasks/create_adlsgen2.yml

```diff
@@ -0,0 +1,235 @@
+
+- name: Update instance_volumes_preferred  in muchos.props
+  lineinfile:
+    path: "{{ deploy_path }}/conf/muchos.props"
+    regexp: '^instance_volumes_preferred\s*=\s*|^[#]instance_volumes_preferred\s*=\s*'
+    line: "instance_volumes_preferred = {{ InstanceVolumes|join(',') }}"
+
+# Not registering variable because  storage values are not visible immediately
+- name: Create ADLS Gen2 storage acount using REST API
+  azure_rm_resource:
+    resource_group: "{{ resource_group }}"
+    provider: Storage
+    resource_type: storageAccounts
+    resource_name: "{{ item.split('@')[1].split('.')[0] }}"
+    api_version: '2019-04-01'
+    idempotency: yes
+    state: present
+    body:
+      sku:
+         name: "{{ adls_storage_type }}"
+      kind: StorageV2
+      properties:
+         isHnsEnabled: yes
+      location:  "{{ location }}"
+  loop:
+      "{{ InstanceVolumes }}"
+
+# Creating User Assigned identity with vmss_name suffixed by ua-msi if not specified in muchos.props
+# Not registering variable because user identity values are not visible immediately
+- name: Create User Assigned Identity 
+  azure_rm_resource:
+    resource_group: "{{ resource_group }}"
+    provider: ManagedIdentity
+    resource_type: userAssignedIdentities
+    resource_name: "{{ user_assigned_identity if user_assigned_identity !='' else vmss_name + '-ua-msi' }}"
+    api_version: '2018-11-30'
+    idempotency: yes
+    state: present
+    body:
+      location:  "{{ location }}"
+
+# Retrieving  facts about User Assigned Identity 
+- name: Get facts for User Assigned Identity
+  azure_rm_resource_facts:
+    resource_group: "{{ resource_group }}"
+    provider: ManagedIdentity
+    resource_type: userAssignedIdentities
+    resource_name: "{{ user_assigned_identity if user_assigned_identity !='' else vmss_name + '-ua-msi' }}"
+    api_version: '2018-11-30'
+  register: UserAssignedIdentityInfo
+  retries: 20
+  delay: 15
+  until:  UserAssignedIdentityInfo.response|map(attribute='properties')|map(attribute='principalId')|join('') is defined
+
+- name: Update principal_id in muchos.props
+  lineinfile:
+    path: "{{ deploy_path }}/conf/muchos.props"
+    regexp: '^principal_id\s*=\s*|^[#]principal_id\s*=\s*'
+    line: "principal_id = {{ UserAssignedIdentityInfo.response|map(attribute='properties')|map(attribute='principalId')|join('') }}"
+
+# This will be used to assign the MSI for VMSS
+- name: Format User Assigned Identity for API
+  set_fact:
+    UserAssignedIdentityArr: "{{ UserAssignedIdentityInfo.response|default({})|map(attribute='id')|map('regex_replace','^(.*)$','{\"\\1\":{}}')|list}}"
+
+# Retrieve facts about role assignment
+- name: Get role definition id for "Storage Blob Data Contributor"
+  azure_rm_resource_facts:
+    resource_group: "{{ resource_group }}"
+    provider: Authorization
+    resource_type:  roleDefinitions
+    resource_name:  ba92f5b4-2d11-453d-a403-e96b0029c9fe
```

Comment:
This is the Role Id for built-in roles "Storage Blob  Data Contributor" that provides read, write, and delete permission for Azure Storage containers and blobs.  [Built in roles] (https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles)

Reply:
Thanks for the explanation @shannawaz.  Could put that link in a comment, but not sure if thats useful since there is a lot of other Azure knowledge that is needed to understand this.